`Markers`: allow setting `symbol` on individual basis

brisvag opened the pull request:
This PR allows to provide `symbol` on a per-point basis in a `Markers` visual.

Currently, the single `symbol` attribute forces users to create multiple visuals if they want multiple symbols; instead, we can accept an array of symbols and handle everything in the shader. I changed the example `point_cloud.py` to take advantage of this, and everything seems to be working great!

![image](https://user-images.githubusercontent.com/23482191/180774959-9fe75ffa-c6db-4536-afd9-715388935aff.png)


Since we cannot pass enums/strings/integers to the shader, we're forced to convert to float and do some gymnastic to get everything to work, which makes things a bit less readable. One downside of this is that `Marker._data['a_symbol']` will contain a bunch of uninterpretable floats, but since this is a private attribute it shouldn't be a big deal.

brisvag commented on vispy/visuals/markers.py:
```diff
@@ -197,351 +202,260 @@
-    float hbar = max(r3,r4);
-    return min(vbar, hbar);
+# combine all the symbol shaders in a big if-else statement
+symbol_func = f"""
+float symbol(vec2 pointcoord, float size, int symbol) {{
+   {' else'.join(
+    f''' if (symbol == {i}) {{
+        // {name}
+        {shader}
+    }}'''
+    for i, (name, shader) in enumerate(symbol_shaders.items())
+    )}
+}}"""
+
+# aliases
+symbol_aliases = {
+    'o': 'disc',
+    '+': 'cross',
+    '++': 'cross_lines',
+    's': 'square',
+    '-': 'hbar',
+    '|': 'vbar',
+    '->': 'tailed_arrow',
+    '>': 'arrow',
+    '^': 'triangle_up',
+    'v': 'triangle_down',
+    '*': 'star',
 }
-"""
+
+symbol_shader_values = {name: i for i, name in enumerate(symbol_shaders)}
+symbol_shader_values.update({
+    **{alias: symbol_shader_values[name] for alias, name in symbol_aliases.items()},
+})
```
This is where the real changes are: an autogenerated if-else chain that dispatches to the right symbol. The big diff above this is only moving code from isnide individual functions to a dictionary of snippets, but the code itself is identical.

brisvag commented on vispy/visuals/markers.py:
```diff
@@ -716,7 +600,8 @@ def set_data(self, pos=None, size=10., edge_width=1., edge_width_rel=None,
                                       ('a_size', np.float32),
-                                      ('a_edgewidth', np.float32)])
+                                      ('a_edgewidth', np.float32),
+                                      ('a_symbol', np.float32)])
```
Unfortunately we have to use float and not ints, because shaders don't accept ints as attributes and varyings. I'm not sure why, but using `float16` resulted in weird visual bugs (no markers, just flat squares); I suspect it's because there is no implicit conversion?

djhoese reviewed (changes requested):
Wow, nice job. This could be so useful when it comes to optimizing certain visualizations. There are probably multiple open issues requesting something like this (just a guess). From reading the title I initially thought that the GLSL `if` block was going to always include all symbols which really worried me about the performance, but it looks like you've handled that just fine by generating it dynamically.

I have two requests:

1. Could you put the per-marker shader code back in the global namespace and then use those variables in the dictionary you create? In addition to the PR diff being cleaner, I think it makes it easier to read and understand the point of the dictionary and dynamic shader stuff separate from what the individual shaders are doing.
2. Can we please add the `symbol` keyword argument to `__init__` and re-add the property? I don't see why this can't exist in addition to the `set_data` keyword argument. And just to be sure, `symbol` can be a scalar string or an array right? No matter what is done, the `__init__` docstring has to be updated for what `symbol` can be.

brisvag commented:
Diff shoudl be better now. Also, note that I removed the shaders for `rect` and `ellipse`:

they were actually unused (and broken) because they require a tuple as `a_size`, which was impossible to pass. They must have been there for a long time, but unused.

djhoese commented:
> Actually, that's how it is. Unfortunately, there's no way around it

Ah ok. My biggest concern is that performance of the MarkersVisual then becomes dependent on the number of symbols we support. You are right it probably won't make a difference with so few options right now, but I'm always nervous/cautious of `if` blocks like this after dealing with our old viridis colormap implementation (it didn't even work on some GPUs because there were too many branches).

>>    Can we please add the symbol keyword argument to init and re-add the property?
>
> I was being consistent with the rest of the args: if they are properties, they're in __init__, otherwise they get passed directly to set_data. I'm happy to change if you think that's better, but then I should do it with pos, size and all the others.

Oh that's unfortunate. I see your point. So it isn't actually possible (and never was) to create a MarkersVisual without calling `.set_data`? That seems...annoying. I think I'd rather have backwards compatibility in this case and then add the other properties (and/or `__init__` kwargs) in a future PR.

brisvag commented:
> So it isn't actually possible (and never was) to create a MarkersVisual without calling `.set_data`? That seems...annoying.

Actually, if no "data" kwargs are passed, set data is not called, so that works fine :)

> You are right it probably won't make a difference with so few options right now

I'm not *super* sure about this, but my understanding of branching in GPUs tells me that this case (a big `if/else if/else` case) is not that bad; maybe you refer to nested branches? like:

```
if ...
	if... 
		if...
		else
	else
else
```

maybe someone more familiar with the hardware can confirm? @rougier @almarklein.

brisvag commented:
I did a quick benchmark: I took the the `point_cloud` example as it is in this PR, multiplied the number of points by 10 and ran it exactly the same on main, minus the `symbol` kwarg.

FPS was identical when moving around the camera, around 20.

brisvag commented:
> 2\. No matter what is done, the `__init__` docstring has to be updated for what `symbol` can be.

Updated the docstring.

Also, I removed the `scaling` kwarg, which was now deprecated for a couple of version already.

djhoese commented:
:man_facepalming: Oh I didn't even see the `**kwargs` being passed to `set_data` from `__init__`. Ok, that looks much better.

And about GPU branching, no I'm not talking about nested branches. I'm talking about straight `if/elif/elif/elif/elif/elif/elif/elif/else` cases. If you have enough `elif` branches on older GPUs they will just not compile the shader. The GPU doesn't like the non-vector-y way things are being done. Modern GPUs usually have a pretty high value for this. Before vispy colormaps could use textures they were implemented with these if/elif/else statements and the raw viridis colormap didn't actually work and had to be reduced. We don't need to worry about it here.

brisvag commented:
This should fix the test. Yes, I think this could be merged :)

Note for the future: the new `setter` for `symbol` is actually pretty useful in general. I think it would be worth it to implement a similar thing for the other arguments of `set_data` (and the same can be said for other visuals). Not in this PR, but worth discussing.

djhoese commented:
Yes, it is something that needs to be made consistent across Visuals. Thanks.

psobolewskiPhD commented on vispy/visuals/markers.py:
```diff
@@ -651,7 +557,7 @@ def __init__(self, symbol='o', scaling=False, alpha=1, antialias=1, spherical=Fa
 
     def set_data(self, pos=None, size=10., edge_width=1., edge_width_rel=None,
                  edge_color='black', face_color='white',
-                 symbol=None, scaling=None):
```
@brisvag Was removing `scaling` kwarg here intentional?
It breaks the example here for the macOS OpenGL issue.
https://github.com/vispy/vispy/issues/2078#issuecomment-863586236

brisvag replied:
Yes; the idea was to unify the behaviour such as:
- uniforms are just properties with setters (no data needs to be sent to the GPU in buffers)
- `set_data` only takes stuff that needs to be sent as uniforms

Maybe I should have done it in a separate PR and more explicitly, at the time it just seemed like a small refactor that made things cleaner ^^'
